geopandas/plotting.py

```diff
@@ -582,10 +582,21 @@ def plot_dataframe(
         # set categorical to True for creating the legend
         categorical = True
         binedges = [values[~nan_idx].min()] + binning.bins.tolist()
-        categories = [
-            "{0:.2f} - {1:.2f}".format(binedges[i], binedges[i + 1])
-            for i in range(len(binedges) - 1)
-        ]
+        if legend_kwds is not None and 'labels' in legend_kwds:
+            if len(legend_kwds['labels']) != classification_kwds['k']:
```

Comment:
You should check with the actual number of bins, not `k`. mapclassify has some classifiers without k and sometimes returns smaller number of bins than set. Use `binning.k` to do the check, that should be the final number used.